File: assignments/ruby/say/say_test.rb
require 'minitest/autorun'
require_relative 'say'

class SayTest < MiniTest::Unit::TestCase
  def test_0
    assert_equal 'zero', Say.new(0).in_english
  end

  def test_one
    skip
    assert_equal 'one', Say.new(1).in_english
  end

  def test_14
    skip
    assert_equal 'fourteen', Say.new(14).in_english
  end

  def test_twenty
    skip
    # This really shouldn't be twenty-zero
    assert_equal 'twenty', Say.new(20).in_english
  end

  def test_twenty_two
    skip
    assert_equal 'twenty-two', Say.new(22).in_english
  end

  def test_100
    skip
    assert_equal 'one hundred', Say.new(100).in_english
  end

  def test_120
    skip
    assert_equal 'one hundred twenty', Say.new(120).in_english
  end

  def test_123
    skip
    assert_equal 'one hundred twenty-three', Say.new(123).in_english
  end

  def test_1_thousand
    skip
    assert_equal 'one thousand', Say.new(1000).in_english
  end

  def test_1_thousand_234
    skip
    assert_equal 'one thousand two hundred thirty-four', Say.new(1234).in_english
  end

  def test_1_million
    skip
    assert_equal 'one million', Say.new(10**6).in_english
  end

  def test_1_million_and_some_crumbs
    skip
    assert_equal 'one million two', Say.new(1000002).in_english
  end

  def test_1_million_2_thousand_345
    skip
    expected = 'one million two thousand three hundred forty-five'
    assert_equal expected, Say.new(1002345).in_english
  end

  def test_1_billion
    skip
    assert_equal 'one billion', Say.new(10**9).in_english
  end

  def test_really_big_number
    skip
    expected = 'nine hundred eighty-seven billion '
    expected << 'six hundred fifty-four million '
    expected << 'three hundred twenty-one thousand '
    expected << 'one hundred twenty-three'
    assert_equal expected, Say.new(987654321123).in_english
  end

  def test_lower_bound
    skip
    assert_raises ArgumentError do
      Say.new(-1).in_english
    end
  end

  def test_upper_bound
    skip
    assert_raises ArgumentError do
      Say.new(1000000000000).in_english
    end
  end
end

File: assignments/haskell/say/say_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Say (inEnglish)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
       [ TestList inEnglishTests ]

inEnglishTests :: [Test]
inEnglishTests =
  [ testCase "zero" $
    Just "zero" @=? inEnglish (0::Int)
  , testCase "one" $
    Just "one" @=? inEnglish (1::Integer)
  , testCase "fourteen" $
    Just "fourteen" @=? inEnglish (14::Int)
  , testCase "twenty" $
    Just "twenty" @=? inEnglish (20::Int)
  , testCase "twenty-two" $
    Just "twenty-two" @=? inEnglish (22::Int)
  , testCase "one hundred" $
    Just "one hundred" @=? inEnglish (100::Int)
  , testCase "one hundred twenty-three" $
    Just "one hundred twenty-three" @=? inEnglish (123::Int)
  , testCase "one thousand" $
    Just "one thousand" @=? inEnglish (1000::Int)
  , testCase "one thousand two hundred thirty-four" $
    Just "one thousand two hundred thirty-four" @=? inEnglish (1234::Int)
  , testCase "one million" $
    Just "one million" @=? inEnglish (1000000::Int)
  , testCase "one million two" $
    Just "one million two" @=? inEnglish (1000002::Int)
  , testCase "one million two thousand three hundred forty-five" $
    Just "one million two thousand three hundred forty-five" @=?
    inEnglish (1002345::Int)
  , testCase "one billion" $
    Just "one billion" @=? inEnglish (1000000000::Int)
  , testCase "a big number" $
    Just "nine hundred eighty-seven billion six hundred fifty-four million \
         \three hundred twenty-one thousand one hundred twenty-three" @=?
    inEnglish (987654321123::Integer)
  , testCase "lower bound" $
    Nothing @=? inEnglish (-1::Integer)
  , testCase "upper bound" $
    Nothing @=? inEnglish (1000000000000::Integer)
  ]
